Feature/Skill-Settings-In-GUI Revised API

AIIX opened the pull request:
## Description
**Adds the ability for a skill display its skill configuration parsed from Settings Meta JSON in skill GUI.**
This Feature is a complete refactor over the previous implementation https://github.com/MycroftAI/mycroft-core/pull/2175, and works independently of any enclosure / platform skill.

It provides a cleaner and simpler interface for displaying Active Skills settings in GUI.

Adds two methods for skill authors:
- self.gui.register_settings(): Allows skills authors to register their skill configuration in GUI, Adds additional GUI handler events on registration to handle syncing and applying skill settings in the GUI UI interface.
- self.gui.display_settings(): Generates UI interface with settings as per the sections and fields on the fly as defined in the Settings Meta JSON with sync from settings.json and Displays the active skill settings in GUI.

The helper util class used internally is responsible for performing the following actions:
- Populate: Reads the Settings Meta JSON and skills own Settings dictionary to create a data model with synced values for GUI to parse.
- Fetch: Returns the populated data model.
- Update: Performs a sync on the new and updated values from Settings.JSON when settings are updated on the web interface and updates the data model parsed in GUI. 
- Clear: Clears the previous data model.

## How to test
```
def initialize(self):
    self.gui.register_settings()

def display_settings(self): 
    # Example function to call display settings 
    # Can be registered with a voice intent, or connected to any button click event in GUI.
    self.gui.show_settings()
```

## Contributor license agreement signed?
CLA [x] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

AIIX committed:
Add methods to show, update and apply skill settings in GUI

AIIX committed:
add UI for skill settings

AIIX commented:
Responsive UI for sections is generated on the fly based on the Sections and Fields defined in the Settings Meta:

Horizontal Display Example:
![voip-skill-config](https://user-images.githubusercontent.com/19663666/93578156-99ee8380-f9ba-11ea-846f-25fa45abc0e8.png)

Vertical Display Example:
![voip-skill-config-vertical](https://user-images.githubusercontent.com/19663666/93578328-d0c49980-f9ba-11ea-9d93-a8dd8476a0ab.png)

Single Section Example:
![weather](https://user-images.githubusercontent.com/19663666/93578377-de7a1f00-f9ba-11ea-9242-d17c5d98c8cb.png)

AIIX committed:
Clean up debug

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2698))

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2698))

JarbasAl commented:
love this <3 now we just need the selene issue fixed so it does not overwrite local changes

JarbasAl commented:
i assume we need to call self.gui.update_settings on the skill settings changed callback in core? cant that be automated? somewhere in here https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/skills/mycroft_skill/mycroft_skill.py#L293

other than that the PR looks good, i dont understand QML very well to review that, but the api is nice and simple, i like it!

AIIX commented:
> i assume we need to call self.gui.update_settings on the skill settings changed callback in core? cant that be automated? somewhere in here https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/skills/mycroft_skill/mycroft_skill.py#L293
> 
> other than that the PR looks good, i dont understand QML very well to review that, but the api is nice and simple, i like it!

No, it is already automated, this function is not to be run manually, its automatically called on settings page when "mycroft.skills.settings.changed" event is emitted for the skill.

JarbasAl commented:
sry, missed that in the QML files! was looking for it in python

I think the PR is great then, im ok with merging as is

forslund commented on mycroft/enclosure/gui.py:
```diff
@@ -349,6 +350,67 @@ def show_url(self, url, override_idle=None,
+        self.skill.gui.register_handler(update_handler,
+                                        self.skill.gui.update_settings)
+
+    def display_settings(self, override_idle=True,
```
Should this be named `show_settings` to match the other show_x methods

AIIX replied:
fixed, renamed to show_settings

forslund commented on mycroft/enclosure/gui.py:
```diff
@@ -349,6 +350,67 @@ def show_url(self, url, override_idle=None,
+        self.skill.settings[message.data["setting_key"]] = \
+            message.data["setting_value"]
+
+    def update_settings(self, message):
```
You mentioned in the comments that this shouldn't be run manually, should it be renamed to `_update_settings` to make it "hidden"?

AIIX replied:
fixed moved to _apply_settings, _update_settings to make them hidden

forslund commented on mycroft/util/settings_gui_generator.py:
```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 Mycroft AI Inc.
```
Will this file be of use to other things than the gui? if not it may be better to keep it in the enclosure directory maybe?

AIIX replied:
I think it should remain in utils, this helper class can be used by anyone to generate a synced data model from settingsmeta.json and settings.json

forslund replied:
Fair point.

forslund commented on mycroft/enclosure/gui.py:
```diff
@@ -349,6 +350,67 @@ def show_url(self, url, override_idle=None,
+        self.skill.gui.register_handler(apply_handler,
+                                        self.skill.gui.apply_settings)
+        self.skill.gui.register_handler(update_handler,
+                                        self.skill.gui.update_settings)
```
couldn't these be

```python
        self.register_handler(apply_handler, self.apply_settings)
        self.register_handler(update_handler, self.update_settings)
```

AIIX replied:
fixed

forslund reviewed (commented):
This looks great! I haven't run it yet. but everything points to awesomeness!

I made some tiny code style comments but you may choose to ignore them if they don't make sense :)

AIIX committed:
Fix helper init, clean method naming, add back button method

AIIX commented:
Latest commit also adds missing method in remove_pages() to check if the UI is provided by System and adds a handler/method for the back button to return to the previous page from the configuration page.

devops-mycroft commented:
Voight Kampff Integration Test Failed ([Results](https://reports.mycroft.ai/core/PR-2698))

AIIX commented:
All conversations/issues with the implementation seem to be resolved as per @forslund review, what is the merge status for this PR as per the new roadmap https://github.com/MyacroftAI/mycroft-core/projects/3#card-45932300

AIIX commented:
Also Voight Kampff  failing doesn't seem to relate to any issues with the current commit on the PR

krisgesling commented:
Hey, it sounded like @forslund hadn't had a chance to actually test run it. So want to give it a run on a Mark II prototype before we merge it in. 

The roadmap is still a very early draft, so lots needs to get added to that. I'm also wondering if the column "Longer term" gives the wrong impression. Really this is just to have "next" as the highest priority, and "longer term" being the next highest priority.

forslund commented:
What's the best way to try this?

I tried adding a `self.gui.show_settings()` in the weather skill. It showed the page but it was blank.

forslund commented:
I also just realized that the settingsmeta may either be a json _or_ a yaml file (slowly moving over to the yaml format)

krisgesling commented:
Is the settings registration intensive at all? Can we do this automatically for each Skill if a GUI is connected?

It's looking pretty good on the Mark II as well. A few display things I noticed:
- Each setting should have a `label` property that we use as the "title" of that setting. Currently it seems to be using the `name` which is the variable name used in the Skill.
- The `select` type I think needs some tweaking. Checkout the News Skill as a good example. It seems to be showing the value, rather than the label of the option. 
- Also in the News Skill because it's got a long list of options the block of radio buttons takes up more vertical space, and the name of block is vertically centered on the left. I wonder if this should be top aligned so the settings "hang" if that makes sense.

@dschweppe I'm sure you'll want to see these UI screens too.

krisgesling commented:
I've been thinking through the experience of this a bit more, and I need to do some testing to see how it operates and answer these questions but wanted to get the thoughts out of my head... 

Given we don't currently have two way settings syncing, how do we handle the different scenarios  of users modifying a setting in different locations? Eg on device, then choosing something else in Selene. Does the local setting take precedence, do we track when a change was made and take the most recent, if the Skill is restarted will it pull the remote settings and overwrite local changes?

This becomes two questions - what is our desired behaviour, and how do we make that clear to users so we aren't setting up the expectation of a feature that doesn't currently exist?

forslund commented:
It don't look to intensive. we could possibly do it when the show_settings is first called and then when it has been registered once.

your second question is harder to answer I think. These are my thoughts, (and things here are mostly from memory so they may be wrong)

I believe currently the settings set in the GUI will be dominant until a change is made on home. The problem I think is when core is restarted...then all messages from home is considered changed overwriting all settings.

One improvement if not solution is to store the big settings update (all skill's settings comes in one big json structure) from home to disk on shutdown. So when starting up and receiving settings it can check if these are new.

We might also want to separate gui and remote settings in the skill allowing them to exist side by side possibly adding a meta data field for each variable containing a timestamp for the update so the skill's settings can use the most recent version... Not sure how simple a system like that would be... Might not be needed either if the gui gets updated with new settings from home only if a change has been made.

AIIX commented:
> I also just realized that the settingsmeta may either be a json _or_ a yaml file (slowly moving over to the yaml format)

I will add in the yaml file parsing, is there any mycroft skill using yaml settings that I can use ?

AIIX commented:
> Is the settings registration intensive at all? Can we do this automatically for each Skill if a GUI is connected?

I think it should be easy if the skillGUI class is initialized for every skill, could call the registration to happen there instead of manually for each skill, this is provided we want to always have all skills to display their configuration in GUI.

The manual registration provided the choice up to the skill authors if they wanted to display their skill configuration on the GUI. 
 
> It's looking pretty good on the Mark II as well. A few display things I noticed:
> 
> * Each setting should have a `label` property that we use as the "title" of that setting. Currently it seems to be using the `name` which is the variable name used in the Skill.
> * The `select` type I think needs some tweaking. Checkout the News Skill as a good example. It seems to be showing the value, rather than the label of the option.
> * Also in the News Skill because it's got a long list of options the block of radio buttons takes up more vertical space, and the name of block is vertically centered on the left. I wonder if this should be top aligned so the settings "hang" if that makes sense.

I'll look into Label instead of Name for display key of settings fields, I have tested this against the weather skill, time & date skill, and some more skills by Jarbas. Will test the display against the news skill settings fields, should move selection to a vertical column layout that should be better than the row layout so incase there are any number of options it doesn't get cramped up in the limited width of each card.

> @dschweppe I'm sure you'll want to see these UI screens too.

AIIX commented:
> I've been thinking through the experience of this a bit more, and I need to do some testing to see how it operates and answer these questions but wanted to get the thoughts out of my head...
> 
> Given we don't currently have two way settings syncing, how do we handle the different scenarios of users modifying a setting in different locations? Eg on device, then choosing something else in Selene. Does the local setting take precedence, do we track when a change was made and take the most recent, if the Skill is restarted will it pull the remote settings and overwrite local changes?
> 
> This becomes two questions - what is our desired behaviour, and how do we make that clear to users so we aren't setting up the expectation of a feature that doesn't currently exist?

I think this should currently be limited to just showing skill settings in GUI that are defined in settingsmeta.json

On editing skills settings in GUI:
So currently with the two way settings sync not working, any changes made in the GUI aren't sent to Selene, they seem to however update the settings.json for the target key/value on that skill locally, in the best case scenario any changes made to settings.json should be synced back to Selene, this two way sync used to work fine at least in the previous PR.

On editing skills settings in Selene:
This implementation listens for mycroft.skills.settings.changed and updates the GUI settings page with the values from Selene, this sync works fine.

Since the model is built on syncing settings.json values and listening for the settings changed event with settingsmeta.json any modification made on Selene that get pulled down (at any time) in settings.json will always have precedence over local changes.

krisgesling commented:
Hey I've made a quick demo Skill using `settingsmeta.yaml` that also includes all the different settings types we currently support.
https://github.com/krisgesling/test-skill-settings-skill

AIIX committed:
Add yaml support and fix layouts

AIIX commented:
Added yaml support and also fixed the layouts more for different field types, also should now be using labels instead of name for display key:

 - Moved Input type fields to double rows, since labels can be really long and there is no character cap set on them, having the label and entry field in double columns will end up with the entry field becoming barely visible if label is too long, hence each should be in its own row.

- Added support for more fields like numbers only field as per the settings yaml in the demo skill.

![yaml-settings-1](https://user-images.githubusercontent.com/19663666/94258747-f584b800-ff4a-11ea-8a7c-64d8ee687646.png)
![yaml-settings-2](https://user-images.githubusercontent.com/19663666/94258761-fcabc600-ff4a-11ea-88b8-4f403d222306.png)

- Options type fields dependent of on the number and available options and width of the card should auto layout in columns in relation to the available width, its cleaner this way, than having a single column.

- If the options are limited to just 3 they should up as single rows, because even if the width can try fitting those in a the width card, if option labels are too long they will tend to cut of, so better approach is having them on different rows. 

![yaml-settings-3](https://user-images.githubusercontent.com/19663666/94258771-00d7e380-ff4b-11ea-9d88-dd21aa40faee.png)

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2698))

JarbasAl commented:
any update on this one?

forslund commented on mycroft/util/settings_gui_generator.py:
```diff
@@ -0,0 +1,89 @@
+        """ Create a SettingList Object """
+        self.settings_list = []
+
+    def populate(self, skill_id, settings_file, settings_dict, file_type):
```
Could the `file_type` be dropped and `os.path.splitext(settings_file)` or `pathlib.Path(settings_file).suffix` be used to detect the file_type?

AIIX replied:
replaced with pathlib.Path(settings_file).suffix

AIIX committed:
remove file_type as a parameter

pep8speaks commented:
Hello @AIIX! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`mycroft/util/settings_gui_generator.py`](https://github.com/MycroftAI/mycroft-core/blob/cea4448ad7f675ca81ecba0f524ece7fc6d7c97b/mycroft/util/settings_gui_generator.py):

> [Line 19:1](https://github.com/MycroftAI/mycroft-core/blob/cea4448ad7f675ca81ecba0f524ece7fc6d7c97b/mycroft/util/settings_gui_generator.py#L19): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

AIIX committed:
fix pep8 error

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2698))

devops-mycroft commented:
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2698))

AIIX commented:
> any update on this one?

all issues hopefully as pointed are fixed and clear, if not please let me know the concerns to get this merged

krisgesling commented:
I think we would need to at least merge #2707 before this can go in.

However even then it faces a big UX issue until we add in two way settings sync. How do we let users know that changes on the device won't be synced to the backend? Without some indication that this is the case I think this will be very confusing to users.

Really I don't think we can add it as a feature until we have two-way sync, or some way to inform users about it's behaviour.

JarbasAl commented:
It would make more sense to merge this + https://github.com/MycroftAI/mycroft-core/pull/2734 instead of blocking on a backend side fix that has no ETA. 

Don't block a solution without providing another....

forslund commented:
I agree with @JarbasAI, this together with #2734 is also a good pairing. Users could switch to just use local settings.

I also think #2707 should be merged, it makes updating more sane in my opinion.

forslund commented:
Closing PR since we're archiving the repo
